File: typescript-examples/e-commerce-auth-scrapingcourse/api/list.ts
import { BrowserContext, Page } from "playwright";
import { extendPayload } from "@intuned/runtime";
import { goToUrl } from "@intuned/browser";

import { ListInputSchema, listSchema } from "../utils/typeAndSchemas";

type Params = ListInputSchema;

interface Product {
  name: string;
  detailsUrl: string;
}

async function extractProductsFromPage(page: Page): Promise<Product[]> {
  // Wait for the product grid to be visible
  // This ensures the page has loaded before we try to extract data
  const productGrid = page.locator("#product-grid");
  await productGrid.waitFor({ state: "visible" });

  // Find all product items within the grid
  // Each product is contained in a div with class "product-item"
  const productItems = await productGrid.locator(".product-item").all();

  // Array to store all extracted product data
  const products: Product[] = [];

  for (const productItem of productItems) {
    try {
      // Extract the product name
      const nameElement = productItem.locator(".product-name");
      const name = await nameElement.textContent();

      // Extract the product detail page URL
      const linkElement = productItem.locator("a");
      const productUrl = await linkElement.getAttribute("href");

      // Only add the product if all required fields were found
      if (name && productUrl) {
        const product = {
          name: name.trim(),
          detailsUrl: productUrl.trim(),
        };

        products.push(product);
        // extend the payload to trigger the details API
        extendPayload({
          api: "details",
          parameters: product,
        });
      }
    } catch (error) {
      // If extraction fails for a single product, log it but continue with others
      console.error("Failed to extract product data:", error);
      continue;
    }
  }

  return products;
}

export default async function handler(
  params: Params,
  page: Page,
  context: BrowserContext
): Promise<Product[]> {
  // Validate input parameters using schema
  const validatedParams = listSchema.parse(params);

  // Navigate to the dashboard page
  await goToUrl({
    page: page,
    url: "https://www.scrapingcourse.com/dashboard",
  });

  // Extract all product details from the page
  const products = await extractProductsFromPage(page);

  // Return all extracted products
  return products;
}

File: python-examples/e-commerce-auth-scrapingcourse/api/list.py
from playwright.async_api import Page
from typing import TypedDict, List, Optional
from runtime_helpers import extend_payload
from intuned_browser import go_to_url


class Product(TypedDict):
    name: str
    detailsUrl: str


async def extract_products_from_page(page: Page) -> List[Product]:
    # Wait for the product grid to be visible
    # This ensures the page has loaded before we try to extract data
    product_grid = page.locator("#product-grid")
    await product_grid.wait_for(state="visible")

    # Find all product items within the grid
    # Each product is contained in a div with class "product-item"
    product_items = await product_grid.locator(".product-item").all()

    # Array to store all extracted product data
    products: List[Product] = []

    for product_item in product_items:
        try:
            # Extract the product name
            name_element = product_item.locator(".product-name")
            name = await name_element.text_content()

            # Extract the product detail page URL
            link_element = product_item.locator("a")
            product_url = await link_element.get_attribute("href")

            # Only add the product if all required fields were found
            if name and product_url:
                product: Product = {
                    "name": name.strip(),
                    "detailsUrl": product_url.strip(),
                }

                products.append(product)
                # extend the payload to trigger the details API
                extend_payload(
                    {
                        "api": "details",
                        "parameters": product,
                    }
                )
        except Exception as error:
            # If extraction fails for a single product, log it but continue with others
            print(f"Failed to extract product data: {error}")
            continue

    return products


async def automation(
    page: Page,
    params: Optional[dict] = None,
    **_kwargs,
) -> List[Product]:
    # Validate input parameters using schema
    if params is None:
        params = {}

    # Navigate to the dashboard page
    await go_to_url(
        page=page,
        url="https://www.scrapingcourse.com/dashboard",
    )

    # Extract all product details from the page
    products = await extract_products_from_page(page)

    # Return all extracted products
    return products
